lints/lint_mp_allowed_eku.go

```diff
@@ -0,0 +1,74 @@
+/*
+ * ZLint Copyright 2019 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/********************************************************************
+Section 5.3 - Intermediate Certificates
+Intermediate certificates created after January 1, 2019, with the exception
+of cross-certificates that share a private key with a corresponding root
+certificate: MUST contain an EKU extension; and, MUST NOT include the
+anyExtendedKeyUsage KeyPurposeId; and, * MUST NOT include both the
+id-kp-serverAuth and id-kp-emailProtection KeyPurposeIds in the same
+certificate.
+Note that the lint cannot distinguish cross-certificates from other
+intermediates.
+********************************************************************/
+
+package lints
+
+import (
+	"time"
+
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/util"
+)
+
+func hasEKU(cert *x509.Certificate, eku x509.ExtKeyUsage) bool {
+	for _, currentEku := range cert.ExtKeyUsage {
+		if currentEku == eku {
+			return true
+		}
+	}
+
+	return false
+}
+
+type allowedEKU struct{}
+
+func (l *allowedEKU) Initialize() error {
+	return nil
+}
+
+func (l *allowedEKU) CheckApplies(c *x509.Certificate) bool {
+	return util.IsSubCA(c)
+}
+
+func (l *allowedEKU) Execute(c *x509.Certificate) *LintResult {
+	if len(c.ExtKeyUsage) == 0 || hasEKU(c, x509.ExtKeyUsageAny) ||
+		(hasEKU(c, x509.ExtKeyUsageEmailProtection) && hasEKU(c, x509.ExtKeyUsageServerAuth)) {
+		return &LintResult{Status: Error}
+	}
+
+	return &LintResult{Status: Pass}
+}
+
+func init() {
+	RegisterLint(&Lint{
+		Name:          "e_mp_allowed_eku",
```

Comment:
We're generally pretty conservative about what can trigger an error. We should be certain that there is an error present as many CAs will block issuance when they encounter an error. As such, I think that this lint can be at max an informational.

Reply:
I could create a small script that automatically fetches all trusted roots from mozilla (http://mxr.mozilla.org/mozilla/source/security/nss/lib/ckfw/builtins/certdata.txt) and creates a valid go file with the spkis. This would allow us to identify _cross-certificates that share a private key with a corresponding root certificate_. This approach would help block issuance of subs and at the same time allow cross-certificates to be signed. However, it requires manual execution of the script when the mozilla trust store is updated.

Reply:
@fotisl It looks like you landed on implementing a command to do this in this branch. I'd personally prefer to see the lint implemented to be informational to start and have the`cmd/zlint-mozilla-trusted-roots-update` implementation included in a separate follow-up PR. @zakird Do you have a pref here?